cmd/snap/cmd_run.go

```diff
@@ -190,30 +210,72 @@ func maybeWaitForSecurityProfileRegeneration(cli *client.Client) error {
+	logger.Debugf("my system key: %v", my)
 
-	for i := 0; i < timeout; i++ {
-		// TODO: we could also check cli.Maintenance() here too in case snapd is
-		// down semi-permanently for a refresh, but what message do we show to
-		// the user or what do we do if we know snapd is down for maintenance?
-		if _, err := cli.SysInfo(); err == nil {
-			return nil
+	var rspError *client.Error
+
+	for i := 0; i < retryCount; i++ {
+		// refreshes the state of snapd maintenance internally
+		act, err := cli.SystemKeyMismatchAdvice(my)
+		if err == nil {
+			// we have an explicit response from snapd
+			switch act.SuggestedAction {
+			case client.MismatchActionProceed:
+				// we're good to go
+				logger.Debugf("continue execution")
+				return "", nil
+			case client.MismatchActionWaitForChange:
+				// snapd sent us an ID of a change to wait for
+				logger.Debugf("snapd started a regenerate profiles change with ID: %v", act.ChangeID)
+				return act.ChangeID, nil
+			default:
+				return "", fmt.Errorf("internal error: unexpected advice: %q", act.SuggestedAction)
+			}
+		}
+
+		// an error, which could be anything from socket being down, to
+		// relevant method, or our system-key version not being supported by the
+		// API if we're talking to an older snapd, or snapd is simply updating
+		restarting := false
+		if maintErr, ok := cli.Maintenance().(*client.Error); ok && maintErr.Kind == client.ErrorKindDaemonRestart {
+			restarting = true
+		}
+
+		if !restarting && errors.Is(err, client.ErrMismatchAdviceUnsupported) {
+			// snapd does not support system-key mismatch resolution, but it is
+			// responding, so we either woke it up or it was already running. In
+			// the context of home direcotires on NFS-like mounts, if we woke up
+			// snapd, the profiles would have regenerated and thus would account
+			// for our network mounted home, but if snapd was already running
+			// when we poked it and the security profiles may or may not be up
+			// to date, but since we're misisng an API there is no way for us to
+			// tell snapd to do the update
+			logger.Debugf("system-key mismatch advice not supported by snapd")
+			return "", nil
+		} else if errors.As(err, &rspError) {
+			// actual response error
+			if !restarting || rspError.Kind != client.ErrorKindSystemKeyVersionUnsupported {
+				logger.Debugf("ignoring system-key mismatch error: %v", err)
+				return "", nil
+			}
+
+			// edge case when the system-key version we sent is not
+			// supported by snapd, but the daemon appears to be restarting,
+			// hopefully as part of a refresh, and so it makes sense to try again
+			logger.Debugf("system-key version unsupported by snapd, but daemon is restarting")
 		} else {
-			logger.Debugf("cannot obtain system info: %v", err)
+			logger.Debugf("cannot obtain system-key mismatch action: %v", err)
 		}
 		// sleep a little bit for good measure
-		time.Sleep(1 * time.Second)
+		<-timeAfter(1 * time.Second)
```

Comment:
Could we use a time.Ticker instead? I'm somewhat allergic to `time.After` and the leak it can cause.

Reply:
won't ticker be more complicated? you can't really run it outside of the loop as it's ticking on its own cadence, so then using it here locally requires a new ticker each time, which now we must not forget to stop, otherwise they would leak too, won't they? (that sets another trap for someone who may refactor the code, they must not do a silly mistake of deferring ticker's stop)

Reply:
I was thinking about using a ticker like sleep:

```
tckr := time.NewTicker(time.Second)
defer tckr.Stop()
for _ := range <-tckr.C {
}
```

The crucial part is this from `go doc time.NewTicker`:
>  The ticker will adjust the time interval
    or drop ticks to make up for slow receivers. The duration d must be greater
    than zero; if not, NewTicker will panic.

Reply:
This way we would really wait for N seconds (+ any time it took us on the last iteration of the loop).

Reply:
I'll leave a TODO, this will need a wrapper in timeutil and probably a mock in testtime